### ISSUE TITLE
Saving agents

### ISSUE THREAD

**Issue by mctigger** [cpnota/autonomous-learning-library#161]
Currently there seems to be no way to save an agent.

My suggestion is to add `state_dict` to the `Agent` API similar to `torch.nn.Module`. For each agent one can then manually implement this method to for example return the model, optimizer and scheduler states in a single dict. Then a experiment could create checkpoints and resume training after interruption or save the final trained agent.

**Comment by cpnota** [comment]
Trying to come up with a solution for saving/loading models. Saving the entire state of the agent imposes many implementation difficulties, makes the design weird, and produces large files. Instead I'm working on a solution that doesn't require any changes to the core agent code.

The gist of the solution is to create a `factory` object which can be saved/loaded from the disc, which implements methods for construction train/test agents. The factory object stores the necessary models and hyperparameters, but not any internal agent state. The result is usage something like:

```python
# construct the preset
factory = dqn().hyperparameters(lr=1e-3).env(some_env).build()

# run agent in train mode
agent = factory.agent()
for i in range(episodes)
    run_episode(agent, some_env)

# run agent in test mode
test_agent = factory.test_agent()
for i in range(test_episodes)
    run_episode(test_agent, some_env)

# save the model for later
factory.save('dqn.pt')

# load model from disk and watch
factory = torch.load('dqn.pt')
test_agent = factory.test_agent()
for i in range(test_episodes)
    run_episode(test_agent, some_env, render=True)
```

In this design, no changes are needed to the actual `Agent` code. The `TestAgent` becomes a separate object (the `eval()` method gets removed from `Agent`, which I never cared for anyhow), and `all.experiment` and `all.presets` would need significant refactoring.

The result is, I think, fairly elegant. Loading a pretrained agent to watch it, etc., becomes very easy and nice. You cannot fully resume an agent in the middle of training, but you can initialize an agent with a pretrained model and change the hyperparameters. It should be possible to extend this API to build stop/resume functionality on top of the proposed API.
